Allow custom CSS for theme.json elements

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Enables CSS in theme.json styles > elements > CSS
Closes https://github.com/WordPress/gutenberg/issues/48439

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Custom CSS under theme.json styles.elements is not working, even though the schema says it works.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Updates `class WP_Theme_JSON_Gutenberg ` with a condition that checks if there are any items under styles > elements, locates custom css, and adds it to the custom CSS stylesheet.
Updates the global styles output to loop through the elements, locate the CSS, and add the css to the styles array.

## Testing Instructions
Activate Twenty Twenty-Three or Twenty Twenty-Four.
In theme.json, under styles > elements > heading, add the following as the first object:

`
"css": "border:3px solid orange;",
`

Example from TT4:

```
			"heading": {
				"css": "border:3px solid orange;",
				"color": {
					"text": "var(--wp--preset--color--contrast)"
				},
				"typography": {
					"fontFamily": "var(--wp--preset--font-family--heading)",
					"fontWeight": "400",
					"lineHeight": "1.2"
				}
			},
```

Next, open the Site editor, block editor, and the front of the website and confirm that headings have an orange border.


## Screenshots or screencast <!-- if applicable -->

carolinan committed:
Allow CSS under theme.json styles.elements

carolinan commented:
I am taking this out of https://github.com/WordPress/gutenberg/pull/49396 while work continues on the block style variations separately.

aristath commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -1178,6 +1178,17 @@ public function get_custom_css() {
+		// Add the global styles elements CSS.
+		if ( isset( $this->theme_json['styles']['elements'] ) ) {
+			foreach ( $this->theme_json['styles']['elements'] as $element => $node ) {
+				$custom_element_css = _wp_array_get( $this->theme_json, array( 'styles', 'elements', $element, 'css' ) );
```
We're trying to avoid using `_wp_array_get()` whenever possible, so in this case, an `asset()` check would be more appropriate 👍 
```suggestion
				$custom_element_css = isset( $this->theme_json['styles']['elements'][ $element ]['css'] )
					? $this->theme_json['styles']['elements'][ $element ]['css']
					: null;
```

aristath reviewed (commented):
This looks OK to me 👍 

I have a simple improvement suggestion (for performance we should avoid using `wp_array_get()` when the number of args is known), but I don't consider it a blocker 🥇

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan committed:
Update lib/class-wp-theme-json-gutenberg.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan commented:
@aaronrobertshaw @tellthemachines 
Hi
This pull request for custom CSS for theme.json elements is still on the 6.5 editor task board awaiting review.
It adds yet another piece of CSS that needs to be considered with https://github.com/WordPress/gutenberg/pull/57841.
I'm thinking this should be punted?

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

aaronrobertshaw commented:
Thanks for the ping 👍 

> It adds yet another piece of CSS that needs to be considered with https://github.com/WordPress/gutenberg/pull/57841.

If I understand it correctly, the addition of the custom CSS shouldn't be an issue. The selector it is added to the ruleset under though might just need a tweak to match what is settled on in https://github.com/WordPress/gutenberg/pull/57841.

> I'm thinking this should be punted?

The goal is to land https://github.com/WordPress/gutenberg/pull/57841 for 6.5 and this would only need a minor tweak I think if #57841 does get merged, so I don't know that it should be punted just yet. Happy to defer to @tellthemachines though 🙂

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan commented:
https://github.com/WordPress/gutenberg/pull/57841 is not planned for 6.5.

But his seems to have been removed from the 6.5 editor task board without comment, I suppose that confirms it is punted.

carolinan commented:
Correction, the *issue* is moved to the 6.6 board: not the PR:

carolinan committed:
Merge branch 'trunk' into add/theme-json-elements-css

carolinan committed:
Update the elements CSS to use the block_nodes array.

Remove the unused stylesheet variable in get_block_custom_css_nodes() and use the block_nodes array.

carolinan commented:
@tellthemachines @aaronrobertshaw With the changes in https://github.com/WordPress/gutenberg/pull/58991,
I wonder if this still the best way to add the elements custom CSS.
The feature works, but I'm wondering about the code readability.

The inline docs can be update to reflect that it is not only block nodes that are returned; but **for clarity** and perhaps for future proofing, would it be better to use two separate methods?
I mean, introducing a public function for elements separately like `get_element_custom_css_nodes()`.

Or, if they don't need to be separate methods, can we make the if statements neutral to the type, and make one if statement work for both blocks and elements?

tellthemachines commented:
@carolinan that's a great question! Based on @oandregal's [recent feedback ](https://github.com/WordPress/gutenberg/pull/58991/files#r1507927224), I was considering refactoring those functions in `class WP_Theme_JSON_Gutenberg` to try and reuse the same logic for custom CSS as we do for the other global styles. That would mean potentially that whole function would be deprecated. I'm currently at WordCamp Asia and will only be back to code work sometime next week, but will look into it then!

richtabor commented:
@carolinan @tellthemachines what're the remaining blockers for this? Seems like a nice improvement to meet expectations on this front.

aaronrobertshaw commented:
What are the plans or expectations for custom CSS for elements when it comes to the extended block style variations within https://github.com/WordPress/gutenberg/pull/57908?

I expect we'll want them there too but given the complexity of that work already, do we have any leeway to hold off here until that lands?
